src/runtime/vulkan/vulkan.cc

```diff
@@ -990,57 +1058,59 @@ Module VulkanModuleCreate(std::unordered_map<std::string, VulkanShader> smap,
 
 VulkanThreadEntry* VulkanThreadEntry::ThreadLocal() { return VulkanThreadStore::Get(); }
 
-VulkanStagingBuffer* VulkanThreadEntry::StagingBuffer(int device_id, size_t size) {
-  if (!staging_buffers_[device_id]) {
-    staging_buffers_[device_id] = std::unique_ptr<VulkanStagingBuffer>(new VulkanStagingBuffer());
+VulkanHostVisibleBuffer* GetOrAllocate(
```

Comment:
Please document the rationale behind why do we do sync after re-alloc(due to deferred execution) so future maintainers can understand the rationale

Reply:
ok I wrote something but as I wrote, I realized that I didn't really understand this issue deeply. My main confusion point is: if we want to make sure that old tasks get finished before we delete the buffer they use, shouldn't we sync before we call `DeleteHostVisibleBuffer`? Or `DeleteHostVisibleBuffer` itself, that calls `vkFreeMemory` etc, is async and it only records the commands to the queue?

Reply:
@masahi you are right, we should sync before we call DeleteHostVisible buffer, great catch!

Reply:
ok updated. Though at this point I'm not sure the way we sync on staging buffer, namely after realloc + memcpy is done, is correct with respect to this point.

Reply:
In the case of staging buffer the sync happens everytime after the copy, so it should be fine. So there won't be stale staging buffer(because previous copy was synced)